canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -106,23 +109,67 @@ private[apiserver] final class ApiPackageManagementService private (
   private def decodeAndValidate(
       darFile: ByteString
   )(implicit
-      contextualizedErrorLogger: ContextualizedErrorLogger
+      contextualizedErrorLogger: ContextualizedErrorLogger,
+      loggingContext: LoggingContextWithTrace
   ): Future[Dar[Archive]] = Future.delegate {
     // Triggering computation in `executionContext` as caller thread (from netty)
     // should not be busy with heavy computation
-    val result = for {
-      darArchive <- Using(new ZipInputStream(darFile.newInput())) { stream =>
-        darReader.readArchive("package-upload", stream)
-      }
-      dar <- darArchive.handleError(Validation.handleLfArchiveError)
-      packages <- dar.all
-        .traverse(Decode.decodeArchive(_))
-        .handleError(Validation.handleLfArchiveError)
-      _ <- engine
-        .validatePackages(packages.toMap)
-        .handleError(Validation.handleLfEnginePackageError)
+    for {
+      (dar, decodedDar) <- Future.fromTry(
+        for {
+          darArchive <- Using(new ZipInputStream(darFile.newInput())) { stream =>
+            darReader.readArchive("package-upload", stream)
+          }
+          dar <- darArchive.handleError(Validation.handleLfArchiveError)
+          decodedDar <- dar
+            .traverse(Decode.decodeArchive(_))
+            .handleError(Validation.handleLfArchiveError)
+          _ <- engine
+            .validatePackages(decodedDar.all.toMap)
+            .handleError(Validation.handleLfEnginePackageError)
+        } yield (dar, decodedDar)
+      )
+      _ <- validateUpgrade(decodedDar)
     } yield dar
-    Future.fromTry(result)
+  }
+
+  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
+      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    val upgradingPackageId = upgradingPackage.main._1
+    logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
```

Comment:
I'd remove this log line since it's a duplicate of the one on line 180

Reply:
fixed